{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "getwd()\n",
    "setwd('..')\n",
    "getwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2000 cells -> 20min\n",
    "# 5000 cells -> 40min\n",
    "# 10000 cells -> 80min"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: plyr\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: Rtsne\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Loading required package: reshape\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    rename, round_any\n",
      "\n",
      "\n",
      "Loading required package: RColorBrewer\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: ape\n",
      "\n",
      "Loading required package: MASS\n",
      "\n",
      "Loading required package: stringi\n",
      "\n",
      "Loading required package: roxygen2\n",
      "\n",
      "Loading required package: phytools\n",
      "\n",
      "Loading required package: maps\n",
      "\n",
      "\n",
      "Attaching package: ‘maps’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plyr’:\n",
      "\n",
      "    ozone\n",
      "\n",
      "\n",
      "Loading required package: repr\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:reshape’:\n",
      "\n",
      "    expand, rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plyr’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plyr’:\n",
      "\n",
      "    desc\n",
      "\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "\n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Package 'mclust' version 5.4.7\n",
      "Type 'citation(\"mclust\")' for citing this R package in publications.\n",
      "\n",
      "\n",
      "Attaching package: ‘mclust’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:maps’:\n",
      "\n",
      "    map\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Matrix’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:reshape’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(SymSim)\n",
    "library(SingleCellExperiment)\n",
    "library(Seurat)\n",
    "\n",
    "library(mclust)\n",
    "library(scran)\n",
    "library(Matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAACoFBMVEUAAAABAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8RERESEhITExMUFBQVFRUYGBgZ\nGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgqKiorKyss\nLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc5OTk6Ojo7Ozs8PDw9PT0+Pj4/\nPz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpMTExNTU1PT09QUFBSUlJTU1NU\nVFRVVVVWVlZXV1dZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZn\nZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6\nenp8fHx9fX1+fn6BgYGCgoKDg4OFhYWHh4eIiIiJiYmKioqLi4uNjY2Ojo6Pj4+QkJCRkZGS\nkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqcnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSl\npaWmpqanp6eoqKipqamurq6vr6+wsLCxsbGzs7O1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9\nvb2+vr6/v7/AwMDBwcHCwsLExMTFxcXGxsbIyMjJycnKysrLy8vNzc3Ozs7Pz8/Q0NDR0dHS\n0tLU1NTV1dXW1tbX19fY2NjZ2dna2trc3Nze3t7f39/g4ODh4eHj4+Pk5OTl5eXm5ubn5+fp\n6enq6urr6+vs7Ozu7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8\n/Pz9/f3+/v7///86JRXaAAAACXBIWXMAABJ0AAASdAHeZh94AAATtUlEQVR4nO3d+7flZUHH\n8WcuIMooA4qiSWDekCQ0L0OOOhaZWsrNC2IwZSUIaY1WFpmpaRczEkWDMUtN0kxLM0lJEczQ\nLFORvDsz/0rMnKHWggfWejifZ5/v9/u8Xr/s56fvZ3/n7Pda++x91ppyAFi3stFPAJZASBAg\nJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkC\nhAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFB\ngJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAk\nCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKE\nBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGA\nkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgs360n\nlGd3nhASy3dp2fSQzhNCYvH+9Ygzf7R8pe+GkFi8p279zAvKB/puCImle2d52YHXlsv7jgiJ\nhfvGg4776oH3l7P6rmxkSNddcy10dfOBAy8vbz5w4EvlR/q+mDcypNPvtR262nPg+q2P/sFt\nL7bjNt3S9cW8kSGdf/4GjjOIp6x9zLCz/H3XGSGxaO8ot3t91x0hsWS3HH/EBS856OnlhV2H\nhMSSXVxesXb4RDml65CQWLBPb33I/6ydvr1ly7d6LgmJBTuj7L39+MjysZ5LQoIAIUGAkCBA\nSBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIE\nCAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKC\nACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQWLhd\n5Tabj3vGh7quCImFO3bL7t27z3tE2fqpnitCYtluKo89+LDvyeV3e84IiWW7uvziocdzy9t7\nzgiJZbtsLaB/uu8Jt/ScERLLtrOcc/HFFz5ly1Nu7DqzlJA+d/pFcCd79x9TDjnpnbkXW81S\nQnpP2egfGVO094ay47aXxy1/c0J5d+7VVrGUkD5c9ucuxnJcVX7l0OMflud23RESi3ZJueLQ\n4zVlV9cdIbFoO8p1hx4vKb/UdUdILNm+bUd+9+Dju4/o+4cNQmLRri/3271799mnlKPe0XdI\nSCzZlYc++z76lItv6jwkJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFB\ngJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAk\nCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKE\nBAFCggAhQYCQIEBIECAkCBASBAiJxbvx137s/lvv//Q//nbHDSGxcPtffWR55PN+/hn3Lmd0\nXBESC3d++eEPH3z82mXndlwREsv2unLyfx8+fr3jjJBYtP84ast1q9jZyJCe+ZhXxpwrJGpe\nUV60kp2NDOkXHr0r5vRkSPfftp1l2PPw8qHcC+NubGRISdG3dg++9FqW4Yay6Vu5F8bdEFLF\nSX+WuxYb6vPlfqsZElKFkBbjP8sR31/JkJAqhLQcJ5b3HT71DUpIFUJajjeUh3784ON33vkz\nXXeEVCGk5dh/Qdn0hOe/eNd9y3ldd4RUIaQl+evnPOiI7Y866+1f7boipAoh0UpIFUKilZAq\nhEQrIVUIiVZCqhASrYRUISRaCalCSLQSUoWQaCWkCiHRSkgVQqKVkCqERCshVQiJVkKqEBKt\nhFQhJFoJqUJItBJShZBoJaQKIdFKSBVCopWQKoREKyFVCIlWQqoQEq2EVCEkWgmpQki0ElKF\nkGglpAoh0UpIFUKilZAqhEQrIVUIiVZCqhASrYRUISRaCalCSLQSUoWQaCWkCiHRSkgVQqKV\nkCqERCshVQiJVkKqEBKthFQhJFoJqUJItBJShZBoJaQKIdFKSBVCopWQKoREKyFVCIlWQqoQ\nEq2EVCEkWgmpQki0ElKFkGglpAoh0UpIFUKilZAqhEQrIVUIiVZCqhASrYRUISRaCalCSLQS\nUoWQaCWkCiHRSkgVQlqSXaWUTdsedvZfBl8hdyakCiEtybHlot0XnXPapvK0b3RcEVKFkBbk\npnLyocfPnV6e13FGSBVCWpCry9lrhy/dq3y+34yQKoS0IJeV3zt82lne1W9mKSF9oPzJW2Ie\n8MaNvh1idpa/O3w6q/xRv5mlhPSRbSflHPHruSf2hSuvZePcvP+YTbd/xnBmuSb3c72jpYQU\nlXxrd/bW7WycPTeUR97+ozix/Evu53pHQqpIhvSaHblr0e6q8vzDp8+WB3XcEVKFkJbjkvKG\nw6fzym933BFShZCWY0f56NrhzeVEX8iumJAWY9+2Ld88+Pj588v2T/UcElKFkBbj+nL07t0X\nPvdRpTy547exB4RUJaTFuLLc5t4PfNLL/6HzkJAqhEQrIVUIiVZCqhASrYRUISRaCalCSLQS\nUoWQaCWkCiHRSkgVQqKVkCqERCshVQiJVkKqEBKthFQhJFoJqUJItBJShZBoJaQKIdFKSBVC\nopWQKoREKyFVCIlWQqoQEq2EVCEkWgmpQki0ElKFkGglpAoh0UpIFUKilZAqhEQrIVUIiVZC\nqhASrYRUISRaCalCSLQSUoWQaCWkCiHRSkgVQqKVkCqERCshVQiJVkKqEBKthFQhJFoJqUJI\ntBJShZBoJaQKIdFKSBVCopWQKoREKyFVCIlWQqoQEq2EVCEkWgmpQki0ElKFkGglpAoh0UpI\nFUKilZAqhEQrIVUIiVZCqhASrYRUISRaCalCSLQSUoWQaCWkCiHRSkgVQqKVkCqERCshVQiJ\nVkKqENKS7CqH/WbPFSFVCGlJjt38ojUf77kipAohLchN5dRVzAipQkgLcnW5aBUzQqq47wNO\nitl+wkbfzeAuK3++ihkhVbzx9W+JefZjN/puBrez/NR5h9zadUZInSXf2n3z8RdcRJO9+485\n/JndcbmfQ42QOkuG9IVy9ka/MOdm7w1lNb+kCqmzZEhfLDfmLjaIq8ovr2RHSJ0JaWNdUoIf\nwd4NIXUmpI21o/zzSnaE1JmQNtS+bVu/s5IhIXUmpA11fTn68B8I/VvfISF1JqQNdeXtf7G6\nue/XSELqTUhjEFJnQhqDkDoT0hiE1JmQxiCkzoQ0BiF1JqQxCKkzIY1BSJ0JaQxC6kxIYxBS\nZ0Iag5A6E9IYhNSZkMYgpM6ENAYhdSakMQipMyGNQUidCWkMQupMSGMQUmdCGoOQOhPSGITU\nmZDGIKTOhDQGIXUmpDEIqTMhjUFInQlpDELqTEhjEFJnQhqDkDoT0hiE1JmQxiCkzoQ0BiF1\nJqQxCKkzIY1BSJ0JaQxC6kxIYxBSZ0Iag5A6E9IYhNSZkMYgpM6ENAYhdSakMQipMyGNQUid\nCWkMQupMSGMQUmdCGoOQOhPSGITUmZDGIKTOhDQGIXUmpDEIqTMhjUFInQlpDELqTEhjEFJn\nQhqDkDoT0hiE1JmQxiCkzoQ0BiF1JqQxCKkzIY1BSJ0JaQxC6kxIYxBSZ0Iag5A6E9JG21XW\nvKDripA6u/TEy2P2lM9u9O3M0LGbX3TIX3VdEVJnv3H86TGnlk9u9O3Mz03lMauYEdKMeGt3\nD1xdLlrFjJBmREj3wGXlilXMCGlGhHQP7CzPPPgb0sWdZ4Q0I9GQHn2f7QPYs/+Ytc/snpb7\nl6sS0oxEQzrjnGsHcPMN5Yzcv9ndENKMREP6yT25a03YVeVlK9kR0owIqd0l5W0r2RHSjAip\n3Y7y6ZXsCGlGhNRs37ajvr+SISHNiJCaXV8ev5ohIc2IkJpdWXavZkhIMyKk6RLSjAhpuoQ0\nI0KaLiHNiJCmS0gzIqTpEtKMCGm6hDQjQpouIc2IkKZLSDMipOkS0owIabqENCNCmi4hzYiQ\npktIMyKk6RLSjAhpuoQ0I0KaLiHNiJCmS0gzIqTpEtKMCGm6hDQjQpouIc2IkKZLSDMipOkS\n0owIabqENCNCmi4hzYiQpktIMyKk6RLSjAhpuoQ0I0KaLiHNiJCmS0gzIqTpEtKMCGm6hDQj\nQpouIc2IkKZLSDMipOkS0owIabqENCNCmi4hzYiQpktIMyKk6RLSjAhpuoQ0I0KaLiHNiJCm\nS0gzIqTpEtKMCGm6hDQjQpouIc2IkKZLSDMipOkS0owIabqENCNCmi4hzYiQpktIMyKk6RLS\njAhpuoQ0I0KaLiHNiJCmS0gzIqTpEtKMCGm6hDQjXyxvuzbmcZdu9O0sipBm5MubStCZG307\niyKkUXlrFyWkUQ0U0r+/6gnbt25/4qtu7LghpFGNE9LrjioP/rkLzzqpbLmm34iQRjVMSL9a\nfug9+w8ePnbGB/utCGlUo4R0RTn5i4eP+77Xb0ZIoxokpK9s3/KJVewIaVSDhPS68rMr2RHS\nqJIh3fLW3BfFWTc/ruw9+Ayv232bV+fu+E6ENKpkSK/ZvH2i9tynfPngM/ydg19BPyt3x3ci\npFElQ3rXA3PXyvpaKfsOH99UfqvjkJBGNUZI3ynl64ePLyjv7TgkpFGNEdKBU8tfHD49ovxX\nxx0hjWqQkN5ajj/08fe+D256aM8dIY1qkJAOvLRsesKLLzjzgWVrzw/thDSsUUI68N7nHL/1\n6JOf86YvdV0R0qiGCWk1hDQqIUUJaVRCihLSqIQUJaRRCSlKSKMSUpSQRiWkKCGNSkhRQhqV\nkKKENCohRQlpVEKKEtKohBQlpFEJKUpIoxJSlJBGJaQoIY1KSFFCGpWQooQ0KiFFCWlUQooS\n0qiEFCWkUQkpSkijElKUkEYlpCghjUpIUUIalZCihDQqIUUJaVRCihLSqIQUJaRRCSlKSKMS\nUpSQRiWkKCGNSkhRQhqVkKKENCohRQlpVEKKEtKohBQlpFEJKUpIoxJSlJBGJaQoIY1KSFFC\nGpWQooQ0KiFFCWlUQooS0qiEFCWkUQkpSkijElKUkEZ12hNfGfOsYzb6bjackEZ11mm7Yk45\neqPvZsMJifXz1k5IBAhJSAQISUgECElIBAhJSAQISUgECElIBAhJSAQISUgECElIBEw6pF3l\nw2uH3eVP+60IifWbdEjHbrl17fDY8sl+K0Ji/aYc0k3l1LXDN7fe63v9ZoTE+k05pKvLhWuH\nj5Qf7zgjJNZvyiFdVt66dnhteWnHGSGxflMOaWd5xnmHPKxc0XFGSKxfNKSnPv+inL37jyn/\n59PBZ3lHQmL9oiFt/elkSDeUJ65d9qvlPj8IPss7EhLrFw3pqPcFL3bgqtt/M/rb8qTkde9I\nSKzfhEO65P8/a3hZ8rp3JCTWb8Ih7bj9W9hzytuT170jIbF+0w1p37Yjv7t2ekT5TPC6dyIk\n1m+6IV1fTls73Lp5277gde9ESKzfdEO6srxk7fDR8hPBy96ZkFi/6Ya0MkJi/YQkJAKEJCQC\nhCQkAoQkJAKEJCQChCQkAoQkJAKEJCQChCQkAoQkJAKEJCQChCQkAoQkJAKEJCQChCQkAoQk\nJAKEJCQChCQkAoQkJAKEJCQChCQkAoQkJAKEJCQChCQkAoQkJAKEJCQChCQkAi4/8vSczT3/\nz+RuhMT63Xj+5Tlb92707dwTQmJivLWDACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBA\nSBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQkA5pV/lQ\n9oJVQmJi0iEdu+nr2QtWCYmJCYd0U3l49Hp3QUhMTDikq8u50evdBSExMeGQLiu/H73eXRAS\nExMOaWf5SPR6d0FITMzmbdtz9uw/ZvOtq3jWQmJirn7/tTk331AetZJnLSQW7arywpXsCIlF\nu6T8wUp2hMSi7Sj/uJIdIbFk+7Zt/dZKhoTEkl1fTl3NkJBYsivLBasZEhIECAkChAQBQoIA\nIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQ\nICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJ\nAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAh\nQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAg\nJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkC\nhAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFB\ngJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAk\nCBASBAgJAv4XryXLfAN9MckAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "phyla_disc <- read.tree('experiments/tree_sameend.txt')\n",
    "#par(mfrow=c(1,2))\n",
    "plot(phyla_disc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_exp <-function(sce,pct){\n",
    "    expsn <- rowMeans(logcounts(sce))\n",
    "    keep <- order(expsn, decreasing = TRUE)[seq_len(pct/100*length(expsn))]\n",
    "    sce_filtered <- sce[keep, ]  \n",
    "    \n",
    "    sce_filtered\n",
    "}\n",
    "\n",
    "save_sce <- function(sce,path,lowdim='TSNE'){\n",
    "    \n",
    "    sizeFactors(sce) <- calculateSumFactors(sce)\n",
    "    \n",
    "    save_data <- Matrix(as.matrix(assay(sce,'counts')),sparse=TRUE)\n",
    "    \n",
    "    writeMM(save_data,sprintf(\"%s.data.counts.mm\",path))\n",
    "    write.table(as.matrix(rownames(save_data)),sprintf('%s.data.row',path),row.names=FALSE, col.names=FALSE)\n",
    "    write.table(as.matrix(colnames(save_data)),sprintf('%s.data.col',path),row.names=FALSE, col.names=FALSE)\n",
    "    \n",
    "    tsne_data <- reducedDim(sce, lowdim)\n",
    "    colnames(tsne_data) <- c(sprintf('%s_1',lowdim),sprintf('%s_2',lowdim))\n",
    "    print(head(cbind(as.matrix(colData(sce)),tsne_data)))\n",
    "    write.table(cbind(as.matrix(colData(sce)),tsne_data),sprintf('%s.metadatacol.tsv',path),row.names=TRUE, col.names=TRUE,sep='\\t')    \n",
    "    write.table(cbind(as.matrix(rowData(sce))),sprintf('%s.metadatarow.tsv',path),row.names=TRUE, col.names=TRUE,sep='\\t')    \n",
    "    \n",
    "    write.table(sizeFactors(sce),file=sprintf('%s.size_factor.tsv',path),sep='\\t',row.names=FALSE, col.names=FALSE)    \n",
    "    \n",
    "}\n",
    "\n",
    "save_hvg_seurat <-function(sce,path,method){\n",
    "    seurat <- CreateSeuratObject(counts = counts(sce), project = \"scRNAseq\", assay = \"RNA\",\n",
    "                                         min.cells = 0, min.features = 0,\n",
    "                                         meta.data = as.data.frame(colData(sce)))\n",
    "    #seurat <- ScaleData(seurat, vars.to.regress = \"nUMI\", display.progress = FALSE)\n",
    "    seurat <- NormalizeData(seurat)\n",
    "    seurat <- FindVariableFeatures(seurat, selection.method = method, nfeatures = length(rownames(seurat@assays$RNA)))\n",
    "    vf <- VariableFeatures(seurat)  \n",
    "    write.table(vf,file=sprintf('%s.hvg.%s.tsv',path,method),sep='\\t',row.names=FALSE, col.names=TRUE)       \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncells_total=10000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  4735, 2910, 4816, 3989, 3252, 4503, 3098, 4824, 4765, 1280 \n",
      "\t   3295, 4686, 3085, 2958, 4224, 4495, 1520, 473, 2271, 1622 \n",
      "\t   814, 1334, 1744, 2064, 2805, 3107, 2523, 3125, 2300, 1472 \n",
      "Negative:  2039, 3090, 942, 132, 2586, 3778, 4741, 3018, 3944, 3816 \n",
      "\t   4092, 2575, 2932, 4934, 2842, 3157, 4135, 4205, 4463, 392 \n",
      "\t   2744, 4995, 125, 3531, 4679, 2686, 3726, 4276, 3541, 2720 \n",
      "PC_ 2 \n",
      "Positive:  2586, 2910, 4503, 3252, 1280, 4735, 4205, 3090, 132, 4135 \n",
      "\t   3810, 4686, 2932, 3098, 4224, 3157, 4495, 3085, 3295, 3989 \n",
      "\t   4435, 4824, 3778, 4765, 4816, 3281, 3018, 3711, 4741, 353 \n",
      "Negative:  1982, 96, 1879, 3573, 3660, 3125, 3107, 4217, 678, 4337 \n",
      "\t   942, 3524, 4636, 1171, 4443, 3944, 2805, 3706, 3365, 2983 \n",
      "\t   4725, 3491, 4998, 4884, 2101, 36, 2744, 3228, 4463, 2897 \n",
      "PC_ 3 \n",
      "Positive:  4636, 678, 1171, 3228, 4509, 3213, 3188, 3095, 3880, 4371 \n",
      "\t   2910, 964, 65, 1848, 2230, 3932, 2972, 2843, 1926, 110 \n",
      "\t   1280, 4135, 1271, 4100, 1823, 4503, 674, 4072, 3252, 4735 \n",
      "Negative:  2586, 3090, 3107, 3125, 2805, 4205, 4443, 2983, 3157, 132 \n",
      "\t   4725, 2932, 3778, 4998, 3706, 4884, 3018, 4934, 3816, 3365 \n",
      "\t   942, 4741, 3491, 1268, 2829, 4124, 2833, 2335, 1002, 952 \n",
      "PC_ 4 \n",
      "Positive:  1879, 3573, 1982, 96, 3660, 4217, 4643, 4612, 4337, 4742 \n",
      "\t   3524, 2101, 4001, 942, 3810, 4975, 353, 4435, 4135, 826 \n",
      "\t   924, 3944, 3516, 2897, 1345, 1551, 2744, 1188, 1349, 4463 \n",
      "Negative:  3107, 3125, 2805, 2983, 4725, 4443, 4998, 4884, 3365, 3491 \n",
      "\t   3706, 2833, 4503, 2829, 4735, 3252, 1280, 952, 4020, 2335 \n",
      "\t   1171, 2481, 678, 4849, 3295, 539, 3085, 2666, 3603, 4686 \n",
      "PC_ 5 \n",
      "Positive:  4643, 4742, 4001, 4975, 678, 1171, 4636, 4509, 3090, 3228 \n",
      "\t   3213, 3188, 3880, 4580, 3095, 4100, 2843, 3171, 4371, 65 \n",
      "\t   1204, 4915, 4072, 942, 3944, 964, 132, 1848, 109, 3816 \n",
      "Negative:  1879, 3573, 1982, 96, 3660, 3810, 4435, 4217, 4337, 3524 \n",
      "\t   353, 826, 3516, 2910, 4864, 1188, 4612, 2897, 3098, 2101 \n",
      "\t   4503, 1280, 271, 3766, 2230, 4735, 924, 4824, 4224, 3932 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 10000\n",
      "Number of edges: 504449\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.7235\n",
      "Number of communities: 36\n",
      "Elapsed time: 1 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1 singletons identified. 35 final clusters.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 0.4008078\n",
      "       orig.ident nCount_RNA nFeature_RNA phenoid RNA_snn_res.2 seurat_clusters\n",
      "cell_1          1      28042         2702       1            31               6\n",
      "cell_2          1      18347         2305       1            23              30\n",
      "cell_3          1      13785         1640       1             8              16\n",
      "cell_4          1      19139         2510       1            31               6\n",
      "cell_5          1      34333         2815       1            31               6\n",
      "cell_6          1       3157          841       1            34               9\n",
      "       ident sizeFactor    TSNE_1      TSNE_2\n",
      "cell_1     6  1.0210994 16.113389   7.3587563\n",
      "cell_2    30  0.6566778 -9.434790   2.5575138\n",
      "cell_3    16  0.3744907 -5.166711   2.4194810\n",
      "cell_4     6  0.8602096 11.841510   8.2351556\n",
      "cell_5     6  1.2378606  9.923945  -0.4352574\n",
      "cell_6     9  0.1213095 17.584796 -10.1662801\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  4276, 4960, 4612, 4432, 2909, 4177, 3136, 3724, 4096, 4188 \n",
      "\t   2989, 1418, 1075, 1947, 3828, 840, 2963, 1339, 2725, 53 \n",
      "\t   1443, 1956, 61, 2648, 4258, 3208, 721, 2024, 1373, 209 \n",
      "Negative:  1246, 3493, 4910, 2749, 2515, 1570, 3260, 4614, 4541, 4926 \n",
      "\t   2848, 3414, 2750, 3922, 3768, 4256, 4352, 4706, 3674, 3247 \n",
      "\t   116, 3932, 2355, 1430, 4687, 3286, 3925, 104, 3502, 2499 \n",
      "PC_ 2 \n",
      "Positive:  2648, 721, 1373, 2718, 1463, 4910, 1570, 2494, 3414, 1246 \n",
      "\t   2515, 770, 1008, 2749, 1430, 1055, 3493, 116, 717, 2355 \n",
      "\t   2499, 3630, 173, 1480, 3674, 1208, 3260, 2179, 2848, 3932 \n",
      "Negative:  4276, 4612, 4960, 4177, 4096, 3724, 2909, 4432, 3136, 4188 \n",
      "\t   3828, 2963, 2989, 2725, 1075, 1947, 840, 1339, 53, 1418 \n",
      "\t   3704, 2528, 3538, 4461, 1792, 185, 3789, 3209, 2772, 2983 \n",
      "PC_ 3 \n",
      "Positive:  2304, 1792, 3538, 3884, 4100, 4138, 2774, 3047, 3012, 4461 \n",
      "\t   2823, 4177, 4498, 3826, 4352, 3674, 3286, 3925, 4612, 2790 \n",
      "\t   1457, 1329, 290, 104, 3932, 235, 2923, 3367, 2664, 2423 \n",
      "Negative:  4910, 3414, 1430, 2061, 2499, 144, 2132, 1223, 4399, 721 \n",
      "\t   2648, 3653, 185, 4437, 4258, 441, 4233, 1373, 2698, 209 \n",
      "\t   2479, 3249, 3516, 3325, 2402, 4504, 3909, 937, 3425, 2089 \n",
      "PC_ 4 \n",
      "Positive:  4352, 3674, 3932, 104, 3286, 3925, 3367, 3476, 3686, 3205 \n",
      "\t   4403, 418, 3060, 4216, 4499, 4258, 4733, 839, 2024, 560 \n",
      "\t   4910, 3414, 2479, 2534, 4119, 960, 209, 10, 1430, 2341 \n",
      "Negative:  1246, 3493, 2923, 2515, 2749, 1570, 3260, 4926, 4614, 4541 \n",
      "\t   2848, 3922, 2750, 3768, 4706, 116, 2355, 4256, 3247, 4687 \n",
      "\t   3502, 4504, 4486, 2857, 4399, 3972, 4612, 173, 1372, 4626 \n",
      "PC_ 5 \n",
      "Positive:  2923, 4352, 3674, 3925, 3286, 104, 3932, 4612, 4276, 4960 \n",
      "\t   1246, 3136, 4432, 3493, 3205, 3367, 3724, 4096, 2749, 3476 \n",
      "\t   2515, 2909, 3686, 4403, 2989, 2848, 3260, 4926, 418, 1570 \n",
      "Negative:  3538, 1792, 4100, 3884, 2304, 4138, 2774, 4461, 4910, 3414 \n",
      "\t   1457, 2499, 235, 1430, 1038, 3852, 4504, 4399, 2823, 3012 \n",
      "\t   799, 4498, 3047, 2664, 3826, 4482, 3721, 3103, 651, 209 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 10000\n",
      "Number of edges: 363318\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.6608\n",
      "Number of communities: 34\n",
      "Elapsed time: 1 seconds\n",
      "[1] 0.3138898\n",
      "       orig.ident nCount_RNA nFeature_RNA phenoid RNA_snn_res.2 seurat_clusters\n",
      "cell_1          1      32418         2935       1             2               2\n",
      "cell_2          1      30023         2916       1            12              12\n",
      "cell_3          1      29459         2865       1             2               2\n",
      "cell_4          1      25327         2694       1             3               3\n",
      "cell_5          1      31325         2749       1             3               3\n",
      "cell_6          1       7589         1456       1             3               3\n",
      "       ident sizeFactor     TSNE_1     TSNE_2\n",
      "cell_1     2  1.4276675   1.941197  -9.023638\n",
      "cell_2    12  1.3011523 -16.015577 -11.684590\n",
      "cell_3     2  1.3672556  -3.465208  -8.376238\n",
      "cell_4     3  1.0873831  -2.966202 -16.721714\n",
      "cell_5     3  1.1995871   4.738862 -15.840950\n",
      "cell_6     3  0.2659608   8.379659  15.014861\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  3515, 3451, 3846, 3655, 4944, 2628, 2769, 525, 4514, 3785 \n",
      "\t   3803, 3676, 4975, 1793, 4985, 4270, 1596, 3007, 3343, 4807 \n",
      "\t   3009, 3521, 3313, 3392, 1574, 759, 1636, 3142, 3052, 4315 \n",
      "Negative:  3948, 4283, 4321, 3498, 2187, 4813, 1963, 4940, 3306, 3568 \n",
      "\t   2437, 4076, 4153, 4693, 1774, 2993, 2773, 2140, 2559, 1154 \n",
      "\t   2382, 1017, 4035, 529, 2327, 4265, 1151, 3181, 1835, 1450 \n",
      "PC_ 2 \n",
      "Positive:  3846, 1793, 4975, 4514, 3451, 4985, 3676, 4807, 4944, 427 \n",
      "\t   4270, 3803, 2769, 3343, 525, 3165, 1191, 508, 3515, 4315 \n",
      "\t   3785, 3655, 4559, 4859, 2859, 2277, 4066, 2529, 759, 1702 \n",
      "Negative:  3392, 1596, 3007, 3009, 3521, 3313, 3142, 3052, 3014, 4216 \n",
      "\t   506, 4561, 3878, 4624, 3836, 1085, 728, 854, 667, 2628 \n",
      "\t   1461, 1783, 1328, 1281, 4891, 3948, 4321, 4283, 3498, 4813 \n",
      "PC_ 3 \n",
      "Positive:  4321, 4283, 3948, 3498, 4153, 4813, 4076, 4940, 4693, 2773 \n",
      "\t   1963, 2437, 1774, 3568, 3306, 3846, 2382, 2993, 529, 1017 \n",
      "\t   1194, 1156, 1450, 656, 1835, 379, 2324, 3451, 2769, 2311 \n",
      "Negative:  4559, 2859, 24, 2711, 3525, 4270, 3199, 4373, 3558, 3181 \n",
      "\t   1793, 3522, 2856, 4315, 4859, 3678, 3343, 3760, 4807, 4372 \n",
      "\t   1052, 4857, 749, 3165, 1577, 2529, 2228, 1596, 818, 3009 \n",
      "PC_ 4 \n",
      "Positive:  1793, 4270, 4807, 3343, 4315, 3165, 4859, 4321, 3498, 4813 \n",
      "\t   4283, 3948, 4076, 4693, 4940, 2711, 24, 2773, 4153, 2277 \n",
      "\t   3525, 3678, 3199, 2529, 2038, 3181, 3760, 4373, 1963, 3522 \n",
      "Negative:  4559, 2859, 3568, 2140, 1151, 427, 1191, 508, 2246, 2461 \n",
      "\t   2559, 4265, 460, 2777, 1187, 1634, 1515, 1289, 2165, 681 \n",
      "\t   1211, 3494, 2364, 4035, 884, 3589, 875, 2104, 810, 483 \n",
      "PC_ 5 \n",
      "Positive:  2628, 3846, 3676, 490, 3343, 2769, 4807, 3803, 3451, 4315 \n",
      "\t   4514, 1793, 525, 4270, 3165, 4985, 3785, 4975, 4944, 3836 \n",
      "\t   3655, 3878, 3392, 2529, 2277, 3678, 4859, 4216, 874, 4561 \n",
      "Negative:  3841, 1995, 2327, 4066, 3732, 3266, 3181, 1555, 2018, 1118 \n",
      "\t   1018, 4058, 24, 869, 1559, 3525, 4332, 2711, 3199, 4373 \n",
      "\t   3522, 3558, 2856, 1577, 749, 2232, 673, 1052, 2567, 2064 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 10000\n",
      "Number of edges: 450782\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.6768\n",
      "Number of communities: 26\n",
      "Elapsed time: 1 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1 singletons identified. 25 final clusters.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 0.5079626\n",
      "       orig.ident nCount_RNA nFeature_RNA phenoid RNA_snn_res.2 seurat_clusters\n",
      "cell_1          1      43622         2693       1             1               1\n",
      "cell_2          1      29914         2083       1            20               5\n",
      "cell_3          1      29006         2277       1            24               9\n",
      "cell_4          1      37589         2332       1            20               5\n",
      "cell_5          1      36300         2511       1             1               1\n",
      "cell_6          1      62642         2903       1            20               5\n",
      "       ident sizeFactor     TSNE_1      TSNE_2\n",
      "cell_1     1  1.1209902  3.0770002   2.5205528\n",
      "cell_2     5  0.6202339 -3.3947945   9.7500518\n",
      "cell_3     9  0.7173706 15.6243213 -28.5199254\n",
      "cell_4     5  0.8093905 -0.4789678   9.8259165\n",
      "cell_5     1  0.9405244  2.5712045   0.8098929\n",
      "cell_6     5  1.4150817 -2.0598860  10.7804525\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  3196, 3588, 2875, 2560, 829, 1278, 2471, 4286, 4246, 4801 \n",
      "\t   3092, 3583, 3752, 3795, 2628, 4290, 3198, 3341, 4909, 3981 \n",
      "\t   3727, 2741, 3849, 2946, 1631, 4427, 4005, 3430, 48, 2656 \n",
      "Negative:  3130, 4910, 2345, 2736, 936, 3024, 4323, 3742, 4979, 4213 \n",
      "\t   4382, 4651, 4857, 4571, 1911, 2895, 402, 4734, 116, 2613 \n",
      "\t   3767, 2304, 2221, 3560, 2097, 1688, 456, 281, 1904, 1163 \n",
      "PC_ 2 \n",
      "Positive:  1081, 2550, 3906, 3044, 3225, 52, 4859, 3554, 4715, 4559 \n",
      "\t   3089, 2593, 4649, 993, 3397, 3450, 972, 2115, 558, 1566 \n",
      "\t   1564, 2255, 1254, 1908, 1981, 4267, 2414, 1493, 2741, 1631 \n",
      "Negative:  4470, 2621, 594, 1278, 2628, 2875, 3092, 2560, 4801, 3583 \n",
      "\t   936, 939, 490, 3742, 4323, 2471, 4213, 4382, 4979, 3024 \n",
      "\t   3752, 3795, 3849, 4005, 84, 4857, 2895, 4734, 969, 2141 \n",
      "PC_ 3 \n",
      "Positive:  3588, 4909, 3198, 3727, 4286, 2946, 3430, 4290, 2974, 2621 \n",
      "\t   4437, 4470, 829, 594, 4246, 4440, 4427, 936, 3341, 3981 \n",
      "\t   3196, 3024, 4728, 4213, 4857, 2895, 4979, 4382, 3742, 4734 \n",
      "Negative:  490, 3130, 3043, 2345, 2736, 1081, 4651, 2741, 4864, 402 \n",
      "\t   1911, 4249, 2550, 116, 1631, 2613, 4571, 3442, 3771, 3906 \n",
      "\t   872, 3225, 3044, 4559, 4842, 2593, 1981, 2097, 3560, 253 \n",
      "PC_ 4 \n",
      "Positive:  3341, 4246, 936, 829, 4290, 3588, 4286, 4910, 4427, 4382 \n",
      "\t   1129, 3198, 3024, 4909, 3742, 4979, 4323, 4213, 3981, 3727 \n",
      "\t   4857, 4728, 5, 2895, 48, 4440, 2946, 3430, 4734, 1648 \n",
      "Negative:  2741, 1631, 3442, 3771, 4842, 2144, 1840, 2875, 2560, 3752 \n",
      "\t   2471, 3092, 3583, 4801, 3795, 3130, 1278, 2345, 4005, 2628 \n",
      "\t   2736, 2656, 3849, 1585, 4855, 4651, 3683, 1666, 116, 1911 \n",
      "PC_ 5 \n",
      "Positive:  2875, 3092, 1081, 3583, 4801, 1278, 2628, 3795, 2560, 3752 \n",
      "\t   2471, 3225, 3044, 3906, 3849, 3089, 2550, 3024, 4005, 3742 \n",
      "\t   4979, 936, 4559, 4213, 3554, 4859, 2656, 4382, 4715, 4323 \n",
      "Negative:  594, 4470, 2621, 2593, 4571, 2741, 1631, 3588, 4290, 4440 \n",
      "\t   2613, 3442, 4246, 3767, 4909, 3771, 829, 4267, 939, 4728 \n",
      "\t   3727, 3341, 4427, 3198, 3430, 2946, 3560, 4437, 4397, 4286 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 10000\n",
      "Number of edges: 500600\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.7190\n",
      "Number of communities: 27\n",
      "Elapsed time: 1 seconds\n",
      "[1] 0.4681857\n",
      "       orig.ident nCount_RNA nFeature_RNA phenoid RNA_snn_res.2 seurat_clusters\n",
      "cell_1          1      29222         2850       1             1               1\n",
      "cell_2          1      24159         2815       1             3               3\n",
      "cell_3          1      21685         2551       1             3               3\n",
      "cell_4          1      29752         2670       1             7               7\n",
      "cell_5          1      19175         2393       1             1               1\n",
      "cell_6          1      24876         2273       1            21              21\n",
      "       ident sizeFactor      TSNE_1     TSNE_2\n",
      "cell_1     1  1.2820358  16.1580772 -0.4609999\n",
      "cell_2     3  1.2476860  -0.6460463 12.7013692\n",
      "cell_3     3  0.9065363  -1.6567045 13.5331326\n",
      "cell_4     7  1.1373606   7.3396861 -2.3190709\n",
      "cell_5     1  0.8910139  15.4450586  1.6598792\n",
      "cell_6    21  0.7267876 -25.1201431 27.3813272\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  2760, 4210, 1917, 4473, 513, 4951, 4363, 4236, 2526, 2543 \n",
      "\t   2046, 2621, 3754, 154, 4910, 3118, 2979, 4277, 4758, 2910 \n",
      "\t   3834, 3576, 1186, 2619, 2776, 2658, 3374, 988, 652, 2029 \n",
      "Negative:  3966, 4580, 2871, 773, 3044, 4187, 2712, 3915, 3099, 4790 \n",
      "\t   4081, 2738, 4320, 1354, 2567, 2274, 2669, 4443, 4327, 1972 \n",
      "\t   4898, 2014, 3708, 68, 3534, 1220, 4403, 3369, 455, 3520 \n",
      "PC_ 2 \n",
      "Positive:  2567, 4320, 2669, 4327, 4898, 3534, 3369, 3708, 3520, 4403 \n",
      "\t   4457, 3775, 622, 4750, 2214, 3756, 2615, 973, 2234, 3776 \n",
      "\t   235, 521, 3014, 4226, 2299, 2116, 1426, 187, 445, 979 \n",
      "Negative:  4324, 4592, 4527, 2624, 4159, 3051, 4147, 2912, 3404, 3557 \n",
      "\t   3313, 4187, 3044, 3980, 773, 4580, 2712, 1354, 2496, 3915 \n",
      "\t   3099, 133, 4790, 4081, 2383, 1969, 2738, 4917, 375, 4362 \n",
      "PC_ 3 \n",
      "Positive:  3708, 4457, 4403, 3775, 3756, 3369, 2669, 4898, 4327, 3520 \n",
      "\t   3534, 4750, 4320, 2615, 3776, 3014, 2567, 4226, 2787, 2751 \n",
      "\t   4860, 3412, 4449, 3157, 4631, 3323, 4370, 2837, 3952, 4855 \n",
      "Negative:  979, 3195, 3521, 401, 1758, 2760, 4951, 4363, 1458, 2526 \n",
      "\t   814, 4236, 2621, 4210, 2046, 2918, 1917, 2543, 607, 1382 \n",
      "\t   4758, 3754, 987, 2031, 4473, 4397, 4910, 1618, 4277, 3118 \n",
      "PC_ 4 \n",
      "Positive:  2740, 4234, 2918, 154, 2046, 2526, 1456, 987, 4951, 4758 \n",
      "\t   2621, 2910, 4236, 4363, 2543, 3754, 2760, 2619, 4910, 4210 \n",
      "\t   1618, 1917, 2979, 3118, 4473, 2031, 4277, 513, 2776, 228 \n",
      "Negative:  1186, 3576, 988, 2658, 3374, 1189, 4069, 3094, 3892, 2704 \n",
      "\t   3564, 3499, 4661, 3406, 1688, 339, 473, 3834, 1344, 651 \n",
      "\t   1200, 2388, 4320, 4467, 2567, 46, 4327, 4457, 2669, 3369 \n",
      "PC_ 5 \n",
      "Positive:  4580, 773, 4081, 3915, 2712, 4790, 2738, 3099, 3044, 4187 \n",
      "\t   1354, 4443, 2014, 1464, 1972, 68, 540, 1576, 1220, 153 \n",
      "\t   1438, 69, 120, 455, 2658, 1186, 1317, 988, 3576, 2918 \n",
      "Negative:  4527, 2624, 4592, 3051, 4159, 3404, 4147, 2912, 4324, 3557 \n",
      "\t   3313, 3980, 2496, 2567, 133, 2669, 2383, 4327, 4898, 1969 \n",
      "\t   4320, 3369, 3708, 3774, 4362, 4903, 3534, 4917, 3520, 1780 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 10000\n",
      "Number of edges: 502744\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.7130\n",
      "Number of communities: 31\n",
      "Elapsed time: 1 seconds\n",
      "[1] 0.4157305\n",
      "       orig.ident nCount_RNA nFeature_RNA phenoid RNA_snn_res.2 seurat_clusters\n",
      "cell_1          1      30920         2546       1             6               6\n",
      "cell_2          1      24139         2418       1             6               6\n",
      "cell_3          1       5045         1336       1             1               1\n",
      "cell_4          1      24319         2611       1             1               1\n",
      "cell_5          1      24960         2492       1             6               6\n",
      "cell_6          1      21902         2071       1             6               6\n",
      "       ident sizeFactor     TSNE_1     TSNE_2\n",
      "cell_1     6  0.9211459 -10.272880  -8.236174\n",
      "cell_2     6  0.8388137  -4.199527 -10.086443\n",
      "cell_3     1  0.2366294 -11.269632 -16.599902\n",
      "cell_4     1  1.0264267 -14.057162 -19.582378\n",
      "cell_5     6  0.8697580 -11.647949 -11.453266\n",
      "cell_6     6  0.6215844 -12.437065  -7.334487\n"
     ]
    }
   ],
   "source": [
    "for(i in c(1:5)){\n",
    "    #############\n",
    "    #True counts\n",
    "    #############\n",
    "    disc_true_counts <- SimulateTrueCounts(ncells_total=ncells_total, \n",
    "                                          min_popsize=50, \n",
    "                                          i_minpop=1, \n",
    "                                          ngenes=5000, \n",
    "                                          nevf=20, \n",
    "                                          evf_type=\"discrete\", \n",
    "                                          n_de_evf=5,\n",
    "                                          vary=\"s\",\n",
    "                                          geffect_mean=1,\n",
    "                                          Sigma=0.4,\n",
    "                                          bimod = 1,\n",
    "                                          phyla=phyla_disc,\n",
    "                                          gene_effect_prob=0.01,\n",
    "                                          randseed=i)\n",
    "    \n",
    "    if(FALSE){\n",
    "        disc_true_counts_tsne <- PlotTsne(meta=disc_true_counts[[3]][,c('cellid','pop')],\n",
    "                                     data=log2(disc_true_counts[[1]]+1),\n",
    "                                     evf_type=\"discrete\",\n",
    "                                     n_pc=30,\n",
    "                                     label='pop',\n",
    "                                     saving = F,\n",
    "                                     plotname=\"discrete populations (true counts)\")\n",
    "        disc_true_counts_tsne[[2]]        \n",
    "    }\n",
    "    \n",
    "    \n",
    "    #############\n",
    "    #Observed counts\n",
    "    #############    \n",
    "    \n",
    "    gene_len <- sample(gene_len_pool, 5000, replace = FALSE)\n",
    "    \n",
    "    \n",
    "    disc_observed_UMI <- True2ObservedCounts(true_counts=disc_true_counts[[1]],\n",
    "                                             meta_cell=disc_true_counts[[3]],\n",
    "                                             protocol=\"UMI\",\n",
    "                                             alpha_mean=0.05,\n",
    "                                             alpha_sd=0.02,\n",
    "                                             gene_len=gene_len,\n",
    "                                             depth_mean=5e5,\n",
    "                                             depth_sd=3e4)   \n",
    "    \n",
    "    if(FALSE){\n",
    "        disc_observed_UMI_tsne <- PlotTsne(meta=disc_observed_UMI[[2]],\n",
    "                                           data=log2(disc_observed_UMI[[1]]+1),\n",
    "                                           evf_type=\"discrete\",\n",
    "                                           n_pc=20,\n",
    "                                           label='pop',\n",
    "                                           saving = F,\n",
    "                                           plotname=\"observed counts UMI\")\n",
    "        disc_observed_UMI_tsne[[2]]           \n",
    "    }\n",
    "    \n",
    "    #############\n",
    "    #Saving\n",
    "    #############    \n",
    "    \n",
    "    \n",
    "    count=disc_observed_UMI[[1]]\n",
    "    colnames(count)=c(disc_observed_UMI[[2]]['cellid'])$cellid\n",
    "    rownames(count)=c(1:dim(count)[1])\n",
    " \n",
    "    seurat <- CreateSeuratObject(counts = count)\n",
    "    seurat@meta.data$phenoid=c(disc_observed_UMI[[2]]['pop'])$pop    \n",
    "    \n",
    "    seurat=FindVariableFeatures(seurat, selection.method = \"vst\", nfeatures = 2000)\n",
    "\n",
    "    seurat_hvg <- NormalizeData(seurat)\n",
    "\n",
    "    all.genes <- rownames(seurat_hvg)\n",
    "    seurat_hvg=ScaleData(seurat_hvg,features = all.genes)\n",
    "\n",
    "    seurat_hvg_pca = RunPCA(seurat_hvg)\n",
    "\n",
    "    PC_cluster=50\n",
    "    PC_2D=50\n",
    "\n",
    "    seurat_hvg_pca <- RunTSNE(seurat_hvg_pca, dims= 1:PC_2D)\n",
    "\n",
    "    seurat_hvg_pca_label <- FindNeighbors(seurat_hvg_pca, dims = 1:PC_cluster)\n",
    "\n",
    "    seurat_hvg_pca_label_cluster <- FindClusters(seurat_hvg_pca_label, resolution = 2)\n",
    "\n",
    "    DimPlot(seurat_hvg_pca_label_cluster,group.by='phenoid',cols = DiscretePalette(22))\n",
    "\n",
    "    DimPlot(seurat_hvg_pca_label_cluster,group.by='seurat_clusters',cols = DiscretePalette(22))\n",
    "\n",
    "    print(adjustedRandIndex(seurat@meta.data$phenoid, seurat_hvg_pca_label_cluster@meta.data$seurat_clusters))\n",
    "\n",
    "    \n",
    "    sce_full_Simul <- as.SingleCellExperiment(seurat_hvg_pca_label_cluster)\n",
    "    \n",
    "    sce_full_Simul_expfiltered=filter_exp(sce_full_Simul,30)\n",
    "    \n",
    "    \n",
    "    path=sprintf('datasets/extract/Simul_%s_1e-2_%s_filtered',ncells_total,i)\n",
    "    \n",
    "    save_sce(sce_full_Simul_expfiltered, path)\n",
    "    saveRDS(disc_true_counts, file = sprintf('datasets/extract/Simul_%s_1e-2_%s_filtered.true_counts.rds',ncells_total,i))\n",
    "    \n",
    "    for(method in c('vst','mvp','disp')){\n",
    "        save_hvg_seurat(sce_full_Simul_expfiltered, path, method)\n",
    "    }    \n",
    "    \n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#help(SimulateTrueCounts)\n",
    "#0.01, 0.005, 0.001, 0.0005 0.0001"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'datasets/extract/Kohinbulk_filtered'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#saveRDS(disc_observed_UMI, file = \"datasets/simul/disc_observed_UMI12.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#saveRDS(disc_true_counts, file = \"datasets/simul/disc_true_counts12.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "'datasets/extract/Kohinbulk_filtered'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'datasets/extract/Simul_1_1e-2_5_filtered'"
      ],
      "text/latex": [
       "'datasets/extract/Simul\\_1\\_1e-2\\_5\\_filtered'"
      ],
      "text/markdown": [
       "'datasets/extract/Simul_1_1e-2_5_filtered'"
      ],
      "text/plain": [
       "[1] \"datasets/extract/Simul_1_1e-2_5_filtered\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“'normalizeSCE' is deprecated.\n",
      "Use 'logNormCounts' instead.\n",
      "See help(\"Deprecated\")”\n",
      "Warning message:\n",
      "“'centreSizeFactors' is deprecated.\n",
      "See help(\"Deprecated\")”\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       orig.ident phenoid    tSNE_1     tSNE_2\n",
      "cell_1          1       1  1.290648 -0.5960434\n",
      "cell_2          1       1  2.295489 -9.3229477\n",
      "cell_3          1       1  2.285761 -0.4733488\n",
      "cell_4          1       1 -3.107209  2.2224019\n",
      "cell_5          1       1 -2.828975  1.0804327\n",
      "cell_6          1       1  1.526465 -0.4216981\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "sce_full_simul=SingleCellExperiment(assays = list(counts = count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "colData(sce_full_simul)$phenoid=c(disc_observed_UMI[[2]]['pop'])$pop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: SingleCellExperiment \n",
       "dim: 500 1000 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(500): 1 2 ... 499 500\n",
       "rowData names(0):\n",
       "colnames(1000): cell_1 cell_2 ... cell_999 cell_1000\n",
       "colData names(0):\n",
       "reducedDimNames(0):\n",
       "spikeNames(0):\n",
       "altExpNames(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sce_full_HumanLiver <- SingleCellExperiment(assays = list(counts = HumanLiverSeurat[['RNA']]@counts))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
